### PR TITLE
chore(block-builder): update block builder to use kafka clients directly

### PR DESCRIPTION
**Pull request by ashwanthgoli**
**What this PR does / why we need it**:

Currently block builders create ephemeral partition reader for processing a new job, but this fails to work as expected because of duplicate kprom metric registration that is triggered on new client creation. We could get around this by wrapping the registerer with `partition_id` and `worker_id` labels and carefully deregistering them when the job processing is complete.

But at this point of time, I prefer directly using the kafka clients (one per each block builder worker) which offers better control over addition/removal of partitions to consume. We can look into partition reader refactor down the line if it's neccessary.

This pr updates block builder to use direct kafka clients, one client per worker. We add partition consume offset when we start a job and remove it at the end of the job to allow reusing the same client.

**Which issue(s) this PR fixes**:
Fixes #<issue number>

**Special notes for your reviewer**:

**Checklist**
- [ ] Reviewed the [`CONTRIBUTING.md`](https://github.com/grafana/loki/blob/main/CONTRIBUTING.md) guide (**required**)
- [ ] Documentation added
- [ ] Tests updated
- [ ] Title matches the required conventional commits format, see [here](https://www.conventionalcommits.org/en/v1.0.0/)
  - **Note** that Promtail is considered to be feature complete, and future development for logs collection will be in [Grafana Alloy](https://github.com/grafana/alloy). As such, `feat` PRs are unlikely to be accepted unless a case can be made for the feature actually being a bug fix to existing behavior.
- [ ] Changes that require user attention or interaction to upgrade are documented in `docs/sources/setup/upgrade/_index.md`
- [ ] If the change is deprecating or removing a configuration option, update the `deprecated-config.yaml` and `deleted-config.yaml` files respectively in the `tools/deprecated-config-checker` directory. [Example PR](https://github.com/grafana/loki/pull/10840/commits/0d4416a4b03739583349934b96f272fb4f685d15)

### CODE REVIEW

File: pkg/blockbuilder/builder/builder.go [hunk 1]
```diff
@@ -501,53 +509,65 @@ func (i *BlockBuilder) processJob(ctx context.Context, job *types.Job, logger lo
...
-			break
+			continue
 		}
 
 		// Reset backoff on successful poll
 		boff.Reset()
 
-		converted := make([]AppendInput, 0, len(records))
-		for _, record := range records {
+		converted := make([]AppendInput, 0, fs.NumRecords())
```
[the remark this answers is not in the document]
Comment (ashwanthgoli) [reply pkg/blockbuilder/builder/builder.go:550]: definitely! i'll take that in a follow-up :)

File: pkg/blockbuilder/builder/builder.go [hunk 2]
```diff
@@ -501,53 +509,65 @@ func (i *BlockBuilder) processJob(ctx context.Context, job *types.Job, logger lo
...
+			level.Error(i.logger).Log("msg", "failed to poll records", "err", err)
 			boff.Wait()
 			continue
 		}
 
-		if len(records) == 0 {
+		if fs.Empty() {
 			// No more records available
-			break
+			continue
```
[the remark this answers is not in the document]
Comment (ashwanthgoli) [reply pkg/blockbuilder/builder/builder.go:544]: this is unlikely, but if we are already caught-up to the latest produced offset this should allow a few retries before failing. I don't have a strong opinion about this change

**Review by owen-d — CHANGES_REQUESTED** [review]
I'm not sure this is a net gain

> We could get around this by wrapping the registerer with partition_id and worker_id labels and carefully deregistering them when the job processing is complete.

We could also just pass the metrics structs to the constructors, which seems like a much less invasive change without tricky downsides.


Unfortunately, passing the client directly as done in this PR:
* ~introduces stateful mutation of offset tracking (`AddConsumePartitions` and cleanup after a job finishes). This will result in some tricky issues if/when we try to process multiple jobs from the same partitions simultaneously~ edit: nvm, this isn't valid since you're creating a client per worker.
* Breaks out of the library we're currently using between both kafka reader impls

edit:
I'm having trouble figuring out where the panic comes from. In main, the readerFactory reuses the same metrics structs which are ultimately initialized in modules.go

If I'm missing something here, feel free to dismiss this review

**Comment by ashwanthgoli** [comment]
> I'm having trouble figuring out where the panic comes from. The readerFactory reuses the same metrics structs which are ultimately initialized once in modules.go

that's right, but we do not register the kprom [metrics](https://github.com/grafana/loki/blob/5bec7d2221e6cc14cf0969057a0f3447d708b753/pkg/kafka/partition/reader.go#L88C3-L88C8) present in this struct unlike the other entries. kprom metrics configured as [client opts](https://github.com/grafana/loki/blob/5bec7d2221e6cc14cf0969057a0f3447d708b753/pkg/kafka/client/writer_client.go#L173) are registered through [OnNewClient](https://github.com/grafana/loki/blob/18c9c05ee127e3c4a40cc885f08fd61a636dfc91/vendor/github.com/twmb/franz-go/plugin/kprom/kprom.go#L126) hook that gets triggered on client creation. This causes duplicate registration panic when we create more than one client

**Review by owen-d — APPROVED** [review]
I see; approved 👍

**Pull request merged**